### PR TITLE
add APM Test Agent to CI for Integration Tests

### PR DESCRIPTION
**Pull request by wconti27**
**What does this PR do?**

This PR adds the APM Test Agent to run alongside integration/plugin tests within CI testing, handling all produced traces during a test run and asserting with any enabled checks. This change was made by modifying the FauxWriter Test Class to emit traces to the test agent.

 The APM Test Agent is being added to every DD APM Tracer with the primary purpose of providing a centralized location for writing tests which will subsequently be used for asserting on handled traces. This strategy will help bring consistency to Datadog APM Tagging/ integration support.  

**Motivation**
We're currently working on integrating the test agent into each tracer's test suites, and for the test agent to handle all traces produced by integrations during testing. Within the Test-Agent, we are then creating checks for Service Naming, Unified Naming Conventions tagging, etc in order to be a centralized location for asserting on integration traces. Allows us to leverage traces produced from integrations during testing.

**How to test the change?**
TO-DO: Verified manually within CircleCI, [such as here](https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/10041/workflows/3309a3cf-2e3c-4bd0-b3a5-619022ac069d/jobs/372744/parallel-runs/5?filterBy=ALL&invite=true#step-10-1556), that integration traces are being emitted to and handled by the Test Agent.

### CODE REVIEW

File: spec/support/faux_writer.rb [hunk 1]
```diff
@@ -20,19 +22,20 @@ def write(trace)
...
       super(trace) if @options[:call_original]
       @traces << trace
+      @options[:transport].send_traces(trace)
     end
   end
 
   def traces(action = :clear)
-    traces = @traces
-    @traces = [] if action == :clear
+    traces = @mutex.synchronize { @traces.dup }
```
Comment (marcotc) [inline spec/support/faux_writer.rb:30]: Why is this mutex necessary?

**Comment by marcotc** [comment]
How does the "APM Test Agent" related to https://github.com/DataDog/system-tests? As far as I understand, system-tests are used to assert consistent behaviours across all tracers, and it does so by capturing data with a testing agent.

File: spec/support/faux_writer.rb (hunk 1 above)
Comment (wconti27) [reply spec/support/faux_writer.rb]: accidental, removed this

**Comment by wconti27** [comment]
> How does the "APM Test Agent" related to https://github.com/DataDog/system-tests? As far as I understand, system-tests are used to assert consistent behaviours across all tracers, and it does so by capturing data with a testing agent.

System Tests is different in the sense that its limited to only testing web frameworks and a select few integrations. What the Test Agent allows is for us to collect traces for all supported integrations across all tracers during testing (the Test Agent is being added at the moment to other tracers' integrations test suites as well), then assert on these traces with checks written centrally within the Test Agent. The Test Agent only has a few checks currently (mainly header checks like trace count and content length), but we'll be expanding it to have checks for Service Naming and Unified Naming for consistent tagging across tracers.

**Comment by marcotc** [comment]
> The Test Agent only has a few checks currently (mainly header checks like trace count and content length), but we'll be expanding it to have checks for Service Naming and Unified Naming for consistent tagging across tracers.

What kind of signal do we get when one of these "checks" fail?

File: .circleci/config.yml [hunk 2]
```diff
@@ -396,7 +436,7 @@ orbs:
         steps:
           - run:
               name: Wait for container on port <<parameters.port>>
```
Comment (GustavoCaso) [inline .circleci/config.yml:438]: We could `<<parameters.host>>` in the name as well

File: docker-compose.yml [hunk 3]
```diff
@@ -186,6 +187,7 @@ services:
     command: /bin/bash
     depends_on:
       - ddagent
+      - testagent
```
Comment (GustavoCaso) [inline docker-compose.yml:190]: Any reason why not using the test agent in all the tracer versions?

File: .circleci/config.yml [hunk 4]
```diff
@@ -176,6 +190,21 @@ step_run_all_tests: &step_run_all_tests
...
+      RESPONSE_CODE=$(echo "$RESPONSE" | awk 'END {print $NF}')
+      if [[ $RESPONSE_CODE -eq 200 ]]; then
+        echo "All Check Traces returned successful! (HTTP 200)"
+      else
+        echo "Check Traces failed with response code: $RESPONSE_CODE"
+        echo "Failures:"
+        cat response.txt
+        exit 1
+      fi
+    when: always
```
Comment (GustavoCaso) [inline .circleci/config.yml:207]: What are the cases when the trace agent considers a trace to have failed? 

Should it make more sense to move this check into the spec suite in ruby?

Comment (wconti27) [reply .circleci/config.yml:207]: > > The Test Agent only has a few checks currently (mainly header checks like trace count and content length), but we'll be expanding it to have checks for Service Naming and Unified Naming for consistent tagging across tracers.
> 
> What kind of signal do we get when one of these "checks" fail?

Hi, please see [here](https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/10224/workflows/60c4cfaa-c497-4ab5-afb8-81e82cebfdcc/jobs/379472#:~:text=Get%20APM%20Test-,Agent,-Trace%20Check%20Results) for the signal, where we query the Test-Agent to see if there were any failing traces submitted. Please keep in mind, there are so many failures since I increased the `X-Datadog-Trace-Count` header by 1 in order to purpos. efully fail the Test-Agent checks, so every received trace is failing.

**Comment by wconti27** [comment]
> > The Test Agent only has a few checks currently (mainly header checks like trace count and content length), but we'll be expanding it to have checks for Service Naming and Unified Naming for consistent tagging across tracers.
> 
> What kind of signal do we get when one of these "checks" fail?

Hi, please see [here](https://app.circleci.com/pipelines/github/DataDog/dd-trace-rb/10224/workflows/60c4cfaa-c497-4ab5-afb8-81e82cebfdcc/jobs/379472#:~:text=Get%20APM%20Test-,Agent,-Trace%20Check%20Results) for the signal, where we query the Test-Agent to see if there were any failing traces submitted. Please keep in mind, there are so many failures since I increased the `X-Datadog-Trace-Count` header by 1 in order to purpos. efully fail the Test-Agent checks, so every received trace is failing.

File: .circleci/config.yml [hunk 5]
```diff
@@ -90,12 +92,24 @@ test_containers:
...
     image: datadog/agent
+    name: ddagent
     environment:
       - DD_APM_ENABLED=true
       - DD_BIND_HOST=0.0.0.0
       - DD_API_KEY=00000000000000000000000000000000
       - DD_HOSTNAME=dd-trace-rb-ci
   - &agent_port 8126
+  - &container_testagent
+    image: williamconti549/dd-apm-test-agent:pool-failures
```
Comment (GustavoCaso) [inline .circleci/config.yml:103]: Will the image be replaced with the official one once we are happy with the implementation? https://github.com/orgs/DataDog/packages?repo_name=dd-apm-test-agent

Comment (wconti27) [reply .circleci/config.yml:103]: Yes it will, just waiting on a PR to be approved in the Test-Agent repository (which this image was produced from). I will switch the images once that PR is merged and released in the Test-Agent

**Comment by wconti27** [comment]
> @wconti27 if you rebase master on your branch, the system tests jobs from CI should be green.
> 
> A part from that do you need any help with this PR?

@GustavoCaso Thank you for the help so far! There is one more piece of functionality I need to add, which is to get the Datadog specific environment at the time of a trace's creation, then to include that environment in the trace headers sent to the Test-Agent. An example would be [this](https://github.com/DataDog/dd-trace-rb/blob/c38d420b0187da4962a5e70386724f936f32388a/spec/datadog/tracing/contrib/service_name_settings_examples.rb#L11) test, where the `DD_TRACE_SPAN_ATTRIBUTE_SCHEMA` env variable is set. Then, within the trace headers: `X-Datadog-Trace-Env-Variables: "DD_TRACE_SPAN_ATTRIBUTE_SCHEMA=v0;DD_OTHER_VAR=VALUE"`

File: .circleci/config.yml (hunk 2 above)
Comment (wconti27) [reply .circleci/config.yml]: Yup, added this

File: spec/support/faux_writer.rb [hunk 6]
```diff
@@ -20,6 +35,7 @@ def write(trace)
     @mutex.synchronize do
       super(trace) if @options[:call_original]
       @traces << trace
+      @options[:transport].send_traces(trace)
```
Comment (GustavoCaso) [inline spec/support/faux_writer.rb:38]: @wconti27 I wonder if we need this line here. 

The call to `super(trace`) will call the transport to send the traces. It depends on which transport we are using. Since it could be the case that `call_original` is set to true, there might be cases where we are sending the traces twice.

File: .circleci/config.yml [hunk 7]
```diff
@@ -271,6 +299,7 @@ orbs:
           - *container_mongo
           - *container_memcached
           - *container_agent
```
Comment (GustavoCaso) [inline .circleci/config.yml:301]: Do we still need to use the `container_agent`?

File: spec/datadog/tracing/contrib/support/tracer_helpers.rb [hunk 8]
```diff
@@ -92,6 +92,11 @@ def clear_traces!
       # with mock assertions.
       config.around do |example|
         example.run.tap do
+          traces = fetch_traces(tracer)
+          traces.each do |trace|
+            # write traces after the test to the agent in order to not mess up assertions
+            tracer.writer.write(trace)
+          end
```
Comment (GustavoCaso) [inline spec/datadog/tracing/contrib/support/tracer_helpers.rb:99]: I'm not sure we want to execute this step all the time, since there might be cases where the agent or the test agent is not running

**Comment by GustavoCaso** [comment]
@wconti27 I don't particularly appreciate that we are hardcoding `'http://testagent:9126'` for `WebMock.

Is there a way to globally configure WebMock in the spec_helper?

**Comment by GustavoCaso** [comment]
I will cc @marcotc on this PR so he can also have a look. Since he is more knowledgeable about the tracing code than me

File: spec/datadog/tracing/contrib/support/tracer_helpers.rb (hunk 8 above)
Comment (wconti27) [reply spec/datadog/tracing/contrib/support/tracer_helpers.rb]: Added a check for this

**Comment by wconti27** [comment]
> @wconti27 I don't particularly appreciate that we are hardcoding `'http://testagent:9126'` for `WebMock.
> 
> Is there a way to globally configure WebMock in the spec_helper?

I wasn't able to find a way to configure WebMock globally, but was able to remove all hardcoded references within each contrib test.

**Comment by wconti27** [comment]
> I will cc @marcotc on this PR so he can also have a look. Since he is more knowledgeable about the tracing code than me

@marcotc cc Can you please review? Thanks!

File: docker-compose.yml [hunk 9]
```diff
@@ -370,9 +385,22 @@ services:
       - "8126"
     ports:
       - "127.0.0.1:${DD_METRIC_AGENT_PORT}:8125/udp"
-      - "127.0.0.1:${DD_TRACE_AGENT_PORT}:8126"
+      - "127.0.0.1:8126:8126"
```
Comment (marcotc) [inline docker-compose.yml:388]: Is this change needed? If so, why?

File: docker-compose.yml [hunk 10]
```diff
@@ -370,9 +385,22 @@ services:
...
     volumes:
       - ddagent_var_run:/var/run/datadog
+
+  testagent:
+    image: ghcr.io/datadog/dd-apm-test-agent/ddapm-test-agent:latest
+    ports:
+        - "127.0.0.1:9126:9126"
```
Comment (marcotc) [inline docker-compose.yml:395]: I suggest creating a new `DD_TEST_AGENT_PORT` variable to configure the host port.

Comment (marcotc) [reply docker-compose.yml:395]:
```suggestion
        - "127.0.0.1:${DD_TEST_AGENT_PORT}:9126"
```

File: docker-compose.yml [hunk 11]
```diff
@@ -370,9 +385,22 @@ services:
...
+    environment:
+        - LOG_LEVEL=DEBUG
+        - TRACE_LANGUAGE=ruby
+        - DD_TRACE_AGENT_URL=http://ddagent:8126
```
Comment (marcotc) [inline docker-compose.yml:399]: Because we reference `ddagent` here, I believe we are missing a
```yaml
depends_on:
  - ddagent
```
for the `testagent` docker service.

File: spec/datadog/tracing/contrib/elasticsearch/transport_spec.rb [hunk 12]
```diff
@@ -10,8 +10,8 @@
 
 RSpec.describe 'Elasticsearch::Transport::Client tracing' do
   before do
-    WebMock.enable!
-    WebMock.disable_net_connect!
+    call_web_mock_function_with_agent_host_exclusions { |options| WebMock.enable! options }
+    call_web_mock_function_with_agent_host_exclusions { |options| WebMock.disable_net_connect! options }
```
Comment (marcotc) [inline spec/datadog/tracing/contrib/elasticsearch/transport_spec.rb:14]: Is `call_web_mock_function_with_agent_host_exclusions` here because one would run this test case without having the agent running locally? If so, that's not a concern as these tests are not meant to successfully run with an agent running.

File: spec/support/faux_writer.rb [hunk 13]
```diff
@@ -1,11 +1,28 @@
...
 # FauxWriter is a dummy writer that buffers spans locally.
 class FauxWriter < Datadog::Tracing::Writer
+  include NetworkHelpers
+
   def initialize(options = {})
-    options[:transport] ||= FauxTransport.new
+    options[:transport] ||= if test_agent_running?
+                              Datadog::Transport::HTTP.default(
+                                headers: {
+                                  'X-Datadog-Trace-Env-Variables' => ENV.to_h.select { |key, _| key.start_with?('DD_') }
```
Comment (marcotc) [inline spec/support/faux_writer.rb:16]: Is `X-Datadog-Trace-Env-Variables` supposed to be an official tracer feature that `dd-trace-rb` does not support yet?

**Comment by marcotc** [comment]
@wconti27, it looks like the `macOS` CI failures are related to the changes in this PR: https://github.com/DataDog/dd-trace-rb/actions/runs/5181100819/jobs/9336151879?pr=2831#step:5:4046, as the request in the stack trace came from `./spec/support/network_helpers.rb:41:in 'check_availability_by_http_request'`

File: spec/support/faux_writer.rb (hunk 13 above)
Comment (wconti27) [reply spec/support/faux_writer.rb:16]: This is not an official tracer feature. The purpose of this is to header is to include the full tracer configuration, which is later used in the Test Agent for assertions (like service-naming by checking the Schema Version used).

Comment (wconti27) [reply spec/support/faux_writer.rb:16]: @marcotc Though on second thought, to achieve this, I believe that I need to include the tracer settings used as well since they can override / be used in addition to the configurable environment variables, correct?

File: spec/datadog/tracing/contrib/elasticsearch/transport_spec.rb (hunk 12 above)
Comment (wconti27) [reply spec/datadog/tracing/contrib/elasticsearch/transport_spec.rb:14]: This is used to allow http web traffic to the test agent, which would otherwise be disabled by webmock for these tests.

File: docker-compose.yml (hunk 9 above)
Comment (wconti27) [reply docker-compose.yml]: removed this change

File: docker-compose.yml (hunk 10 above)
Comment (wconti27) [reply docker-compose.yml]: added this

File: spec/support/faux_writer.rb (hunk 13 above)
Comment (marcotc) [reply spec/support/faux_writer.rb:16]: Hmmm, do we know what tracer configuration we are looking for?

It is correct that `ENV` won't give you the tracer configuration.

I ask because we have a lot of tracer configuration options (I'd guess 100+). We probably don't need them all.

The effective configuration used by the tracer is in the `Datadog.configuration` object (it has a `.to_h` to give you all options, for example)

Comment (wconti27) [reply spec/support/faux_writer.rb:16]: Mainly the configuration that we would like to send over is basic datadog stuff like DD_SERVICE, the span attribute schema version being used (DD_TRACE_SPAN_ATTRIBUTE_SCHEMA), and a few more service naming specific variables.

Comment (marcotc) [reply spec/support/faux_writer.rb:16]: Ok, the best way here is to collect the require data individually from `Datadog.configure` (e.g. `Datadog.configure.service`) and send each one as a header.

Comment (wconti27) [reply spec/support/faux_writer.rb]: great, I made this change!

**Review by marcotc — APPROVED** [review]

**Review by marcotc — APPROVED** [review]
